### PR TITLE
jobs: reduce contention on jobs table by not updating terminal jobs

### PR DESCRIPTION
**Pull request by ajwerner**
Currently our session and claiming logic will interact with terminal jobs.
This is bad both in that claiming may take a very long time because it may
claim already terminal jobs and because expiring out old sessions may end
up updating long ago terminated jobs.

Release note (bug fix): Eliminate opportunity for livelock in jobs subsystem
due to frequent updates to already finished jobs.

### CODE REVIEW

**Comment by spaskob** [comment]
Thanks for looking into this Andy!
Just to make sure I understand - did we ever observe this?
My understanding is that terminal jobs won't be claimed because `claim_session_id IS NULL` will evaluate to false for them, ie the invariant is once a job is claimed, it will alway be claimed by a node.

**Review by ajwerner — COMMENTED** [review]
The problem arises if the node which did the claiming dies. In that case, before this change, the code which would expire sessions would be updating the claims of those terminal jobs back to `NULL`, violating the invariant. I'm updating both to deal with the case of mixed version or scenarios where there are already lots of claimed jobs with `NULL`. 

I observed this recently on a community cluster where there were over 100k completed stats jobs which were completed before a rolling restart of the cluster. It was causing big problems. 

I'm writing some tests now. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56855#-:-MMRD-nb6uqlUCMzeOkc:bncj4mj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spaskob)

**Comment by spaskob** [comment]
Got it - in that case wouldn't it make sense only to change `removeClaimsFromDeadSessions`?

**Review by ajwerner — COMMENTED** [review]
> Got it - in that case wouldn't it make sense only to change removeClaimsFromDeadSessions?

Imagine the cluster I just mentioned. It will have 100k jobs that are terminal will a `NULL` claim. I'd rather add code to be robust to that case than not. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56855#-:-MMRGZFT7pV66njVjsT7:bm4d046)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spaskob)

**Comment by spaskob** [comment]
:lgtm: after test is added

**Review by ajwerner — COMMENTED** [review]
Added testing. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56855#-:-MMRJsT787k282wrYMyA:b35nvml)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @spaskob)

File: pkg/jobs/registry_external_test.go [hunk 1]
```diff
@@ -302,3 +307,96 @@ func TestRegistryResumeActiveLease(t *testing.T) {
...
+   INSERT
+     INTO system.jobs (
+                        status,
+                        payload,
+                        progress,
+                        claim_session_id,
+                        claim_instance_id
+                      )
+   VALUES ($1, $2, $3, $4, $5)
+RETURNING id;
```
Comment (spaskob) [inline pkg/jobs/registry_external_test.go:350]: what is the reason for this alignment of SQL statement?

**Review by spaskob — APPROVED** [review]
LGTM

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56855)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @spaskob)

___
*[pkg/jobs/registry_external_test.go, line 350 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56855#-MMRJrugtXFo6mcgW-oZ-r1-350:-MMRQFo_3PPS1gRlYCAc:b1hlop4) ([raw file](https://github.com/cockroachdb/cockroach/blob/18f06b9aa2730ab6d611d6be9fc67a2a24f8c912/pkg/jobs/registry_external_test.go#L350)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

what is the reason for this alignment of SQL statement?
</blockquote></details>

it's what `sqlfmt` says to do with the line limit set to 80.

**Comment by spaskob** [comment]
Nice, I did not know about sqlfmt!

**Comment by ajwerner** [comment]
TFTR!

bors r=spaskob

**Pull request merged**